pkg/chainsync/chainsync.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package chainsync provides the implementation
+// of the chainsync protocol that verifies peer chain synchronization.
+package chainsync
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/chainsync/pb"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/ratelimit"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/transaction"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+const (
+	protocolName     = "chainsync"
+	protocolVersion  = "1.0.0"
+	syncStreamName   = "prove"
+	messageTimeout   = 1 * time.Minute
+	limitBurst       = 2
+	limitRate        = time.Minute
+	blocksToRemember = 1000
+)
+
+var (
+	errRateLimitExceeded = errors.New("rate limit exceeded")
+)
+
+func (s *ChainSync) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    syncStreamName,
+				Handler: s.syncHandler,
+			},
+		},
+	}
+}
+
+type ChainSync struct {
+	streamer   p2p.Streamer
+	ethClient  transaction.Backend
+	inLimiter  *ratelimit.Limiter
+	outLimiter *ratelimit.Limiter
+	lru        *lru.Cache
+
+	quit chan struct{}
+}
+
+func New(s p2p.Streamer, backend transaction.Backend) (*ChainSync, error) {
+	lruCache, err := lru.New(blocksToRemember)
```

Comment:
It would be great to have caching at the backend so that all parts of the codebase that rely on the backend can get a boost and not have to implement their own local cache. Task for the future.

Reply:
When working on ethproxy, I've witnessed how large the number of calls were to the eth endpoint. We should think of ways of reducing this for network reliability.